### PR TITLE
Feature/genericize network request

### PR DESCRIPTION
**Pull request by erd-s**

### CODE REVIEW

File: ButterLocalServer/main.py [hunk 1]
```diff
@@ -1,26 +1,40 @@
...
 app = FastAPI()
 
 
-@app.get("/test/a")
-def get_test_data_for_endpoint_a():
-    return {"movie_list": ["fast and the furious", "2 fast 2 furious"]}
+# Movies
+@app.get("/movies")
+def get_celeb_movies(name: str):
+	if name == "vin disel":
```
Comment (lanceruss) [inline ButterLocalServer/main.py:14]: You care about the sp?

File: ButterLocalServer/main.py [hunk 2]
```diff
@@ -1,26 +1,40 @@
...
+@app.get("/time")
+def get_time():
+	return {"time": time.time()}
 
 
 if __name__ == "__main__":
-    port = int(os.getenv("PORT", 8070))
-    uvicorn.run(app, host='localhost', port=port)
+	port = int(os.getenv("PORT", 8070))
+	uvicorn.run(app, host='localhost', port=port)
```
Comment (jonpjones-livly) [inline ButterLocalServer/main.py:40]: is indentation in python a codestyling thing or a language requirement?

File: ButterSample/ButterSample/SampleEndpoints/SampleEndpointA.swift [hunk 3]
```diff
@@ -12,11 +12,23 @@ struct SampleEndpointA: Endpoint {
     
     var port: Int? { 8070 }
     
-    var path: String? { "/test/a" }
+    var path: String? { "movies" }
```
Comment (jonpjones-livly) [inline ButterSample/ButterSample/SampleEndpoints/SampleEndpointA.swift:15]: guessing this should be `/movies`

File: ButterSample/ButterSample/SampleEndpoints/SampleEndpointA.swift [hunk 4]
```diff
@@ -12,11 +12,23 @@ struct SampleEndpointA: Endpoint {
...
     
     var httpMethod: HTTPMethod { .get }
     
-    var data: RequestData { .none }
+	var data: RequestData { .params(params: ["name": "vin disel"]) }
     
-    var headers: HTTPHeaders? { ["X-Some-Header": "Some-Header-Argument"] }
+    var headers: HTTPHeaders? { ["ContentType": "application/json"] }
```
Comment (jonpjones-livly) [inline ButterSample/ButterSample/SampleEndpoints/SampleEndpointA.swift:21]: this indentation is weird

File: ButterSample/ButterSample/SampleEndpoints/SampleEndpointB.swift [hunk 5]
```diff
@@ -5,18 +5,36 @@
...
+
 struct SampleEndpointB: Endpoint {
     var scheme: String { "http" }
     
     var host: String { "localhost" }
     
     var port: Int? { 8070 }
     
-    var path: String? { "/test/b" }
+    var path: String? { "bungheads/apply/" }
```
Comment (jonpjones-livly) [inline ButterSample/ButterSample/SampleEndpoints/SampleEndpointB.swift:21]: where is the `/` between the host and path being added? did it change in this diff and i'm not seeing it

File: ButterSample/ButterSample/ViewController.swift [hunk 6]
```diff
@@ -14,46 +14,32 @@ class ViewController: UIViewController {
         super.viewDidLoad()
     }
     
-    func makeRequest(at endpoint: Endpoint) {
+	func makeRequest<T: Decodable>(responseType: T.Type, at endpoint: Endpoint) {
```
Comment (jonpjones-livly) [inline ButterSample/ButterSample/ViewController.swift:17]: indentation is weird

File: ButterSample/ButterSample/ViewController.swift [hunk 7]
```diff
@@ -14,46 +14,32 @@ class ViewController: UIViewController {
...
         let router = Router()
         spinner.startAnimating()
-        router.makeRequest(endpoint: endpoint) { result in
+		let decoder = JSONDecoder()
+		decoder.dateDecodingStrategy = .secondsSince1970
+		router.makeRequest(responseType: responseType, endpoint: endpoint, decoder: decoder) { result in
```
Comment (jonpjones-livly) [inline ButterSample/ButterSample/ViewController.swift:22]: Something we did at nerdery was make the json decoder / encoder a component of the request object (what you've named as your `Endpoint` - reason being that decoding strategies might very well change between different requests, for would be the same for any particular endpoint

File: ButterSample/ButterSample/ViewController.swift [hunk 8]
```diff
@@ -14,46 +14,32 @@ class ViewController: UIViewController {
...
             DispatchQueue.main.async {
                 self.spinner.stopAnimating()
                 switch result {
                 case .success(let data):
-                    self.setOutputTextViewText(jsonData: data)
+					self.setOutputTextViewText(object: data as? CustomStringConvertible)
```
Comment (jonpjones-livly) [inline ButterSample/ButterSample/ViewController.swift:27]: what is going on with your indentation lol

File: ButterSample/ButterSample/ViewController.swift [hunk 9]
```diff
@@ -14,46 +14,32 @@ class ViewController: UIViewController {
...
     }
     
-    func setOutputTextViewText(jsonData: Data?) {
+    func setOutputTextViewText(object: CustomStringConvertible?) {
         outputTextView.textColor = .black
         
-        guard let data = jsonData else {
-            outputTextView.text = "No data"
+        guard let object = object else {
+            self.outputTextView.text = "Successfully retrieved your object. Conform to CustomStringConvertible to print object info."
```
Comment (jonpjones-livly) [inline ButterSample/ButterSample/ViewController.swift:39]: please localize this

Comment (jonpjones-livly) [reply ButterSample/ButterSample/ViewController.swift:39]: <sup>jk</sup>

File: ButterSample/ButterSample/ViewController.swift [hunk 10]
```diff
@@ -70,16 +56,16 @@ class ViewController: UIViewController {
...
     @IBAction func makeRequestB(sender: UIButton) {
         let endpoint = SampleEndpointB()
-        makeRequest(at: endpoint)
+		makeRequest(responseType: SampleResponseB.self, at: endpoint)
     }
     
     @IBAction func makeRequestC(sender: UIButton) {
         let endpoint = SampleEndpointC()
-        makeRequest(at: endpoint)
+		makeRequest(responseType: SampleResponseC.self, at: endpoint)
```
Comment (jonpjones-livly) [inline ButterSample/ButterSample/ViewController.swift:69]: indentation

File: Sources/Butter/Router.swift [hunk 11]
```diff
@@ -4,7 +4,7 @@
 
 import Foundation
 
-public typealias NetworkCompletion = (Result<Data?, Error>) -> ()
+public typealias NetworkCompletion<T: Decodable> = (Result<T, Error>) -> ()
```
Comment (jonpjones-livly) [inline Sources/Butter/Router.swift:7]: what about endpoints that vend an empty success - void is can't conform to decodable afaik

File: Sources/Butter/Router.swift [hunk 12]
```diff
@@ -16,18 +16,22 @@ extension URLSession: URLSessionDataTaskInterface { }
...
+    }
     
-    public func makeRequest(sessionInterface: URLSessionDataTaskInterface? = nil,
-                            endpoint: Endpoint,
-                            completion: @escaping NetworkCompletion) {
-        let session = sessionInterface ?? URLSession.shared
+	public func makeRequest<T: Decodable>(responseType: T.Type,
+										  endpoint: Endpoint,
+										  decoder: JSONDecoder = JSONDecoder(),
+                                          completion: @escaping NetworkCompletion<T>) {
```
Comment (jonpjones-livly) [inline Sources/Butter/Router.swift:28]: indentation

File: Sources/Butter/Router.swift [hunk 13]
```diff
@@ -16,18 +16,22 @@ extension URLSession: URLSessionDataTaskInterface { }
...
         do {
             let requestBuilder = URLRequestBuilder()
             let request = try requestBuilder.request(from: endpoint)
             task = session.dataTask(with: request) { data, response, error in
-                self.handleDataTaskResponse(data: data,
+				self.handleDataTaskResponse(using: decoder,
+											data: data,
                                             response: response,
```
Comment (jonpjones-livly) [inline Sources/Butter/Router.swift:35]: 🤔

File: Sources/Butter/Router.swift [hunk 14]
```diff
@@ -50,8 +55,16 @@ public class Router {
...
+			do {
+				let ret = try decoder.decode(T.self, from: data)
+				completion(.success(ret))
+			} catch {
+				completion(.failure(error))
+			}
+		} else {
+			let error = ButterError.unknown(debugInfo: "No error, no data.")
+			completion(.failure(error))
         }
```
Comment (jonpjones-livly) [inline Sources/Butter/Router.swift:68]: indentation is DOUBLED

File: ButterLocalServer/main.py (hunk 2 above)
Comment (erd-s) [reply ButterLocalServer/main.py:40]: it's a language requirement, i think i opened this in xcode and it messed up this indentation... i will do a separate PR to address indentation across the project

File: ButterLocalServer/main.py (hunk 1 above)
Comment (erd-s) [reply ButterLocalServer/main.py:14]: hah no i did this on purpose

File: ButterSample/ButterSample/SampleEndpoints/SampleEndpointA.swift (hunk 3 above)
Comment (erd-s) [reply ButterSample/ButterSample/SampleEndpoints/SampleEndpointA.swift:15]: i'll check on this.. i think when you do url appending path components it adds in the first "/" for you. this is hitting the local server ok

File: ButterSample/ButterSample/SampleEndpoints/SampleEndpointB.swift (hunk 5 above)
Comment (erd-s) [reply ButterSample/ButterSample/SampleEndpoints/SampleEndpointB.swift:21]: will look into this

File: ButterSample/ButterSample/ViewController.swift (hunk 7 above)
Comment (erd-s) [reply ButterSample/ButterSample/ViewController.swift:22]: i like this idea, i'll update

File: ButterSample/ButterSample/ViewController.swift (hunk 9 above)
Comment (erd-s) [reply ButterSample/ButterSample/ViewController.swift:39]: 😂

File: Sources/Butter/Router.swift (hunk 11 above)
Comment (erd-s) [reply Sources/Butter/Router.swift:7]: i was wondering about that. i'll investigate that

**Comment by erd-s** [comment]
Changed this to go into develop instead of master

File: ButterSample/ButterSample/SampleEndpoints/SampleEndpointA.swift (hunk 3 above)
Comment (erd-s) [reply ButterSample/ButterSample/SampleEndpoints/SampleEndpointA.swift:15]: so it looks like you can do it with or without the leading `/`. so technically it doesn't matter, but it is more readable to have it start with `/` ?

File: Sources/Butter/Router.swift (hunk 11 above)
Comment (erd-s) [reply Sources/Butter/Router.swift:7]: You can do an empty Decodable object here, i'll commit another endpoint that has the example.

Comment (erd-s) [reply Sources/Butter/Router.swift:7]: 94f397d

File: ButterSample/ButterSample/ViewController.swift (hunk 7 above)
Comment (erd-s) [reply ButterSample/ButterSample/ViewController.swift]: updated in f666fef

File: ButterSample/ButterSample/SampleEndpoints/SampleEndpointA.swift (hunk 3 above)
Comment (erd-s) [reply ButterSample/ButterSample/SampleEndpoints/SampleEndpointA.swift:15]: AF leaves it out, i'm gonna leave it out as well.

**Pull request merged**